File: src/file_ops.py
#!/usr/bin/env python3

import os
import shutil
from pathlib import Path
from typing import Optional

from src.types_def import CopyError


def validate_source_path(
    repo_dir: Path, relative_path: str, is_file: bool
) -> Optional[Path]:
    """Validate source path and return Path object if valid"""
    full_path = repo_dir / relative_path

    if not full_path.exists():
        print(f"Warning: {relative_path} not found in repository, skipping...")
        return None

    if is_file and not full_path.is_file():
        print(f"Warning: {relative_path} is not a file, skipping...")
        return None

    if not is_file and not full_path.is_dir():
        print(f"Warning: {relative_path} is not a directory, skipping...")
        return None

    return full_path


def create_symlink(source: Path, target: Path) -> bool:
    """Create a symlink from source to target"""
    if not source.exists():
        raise CopyError(f"Source {source} does not exist")
    try:
        target.parent.mkdir(parents=True, exist_ok=True)
        if target.exists() or target.is_symlink():
            if target.is_dir() and not target.is_symlink():
                shutil.rmtree(target)
            else:
                target.unlink()
        os.symlink(str(source.resolve()), str(target))
        return True
    except Exception as link_error:
        raise CopyError(f"Failed to symlink {source} to {target}: {link_error}")

File: src/file-ops.ts
import { promises as fs } from "node:fs";
import * as path from "node:path";

import { SymlinkError } from "./types-def";

export async function validateSourcePath(
  repoDir: string,
  relativePath: string,
  isFile?: boolean,
): Promise<string | null> {
  const fullPath = path.join(repoDir, relativePath);

  try {
    const stat = await fs.stat(fullPath);

    if (isFile === true && !stat.isFile()) {
      console.warn(`Warning: ${relativePath} is not a file, skipping...`);
      return null;
    }

    if (isFile === false && !stat.isDirectory()) {
      console.warn(`Warning: ${relativePath} is not a directory, skipping...`);
      return null;
    }

    return fullPath;
  } catch {
    console.warn(`Warning: ${relativePath} not found in repository, skipping...`);
    return null;
  }
}

export async function isBrokenSymlink(targetPath: string): Promise<boolean> {
  try {
    const stat = await fs.lstat(targetPath);
    if (!stat.isSymbolicLink()) {
      return false;
    }
    await fs.stat(targetPath);
    return false;
  } catch {
    return true;
  }
}

export async function createSymlink(source: string, target: string): Promise<boolean> {
  try {
    const sourceStat = await fs.stat(source);

    const targetDir = path.dirname(target);
    await fs.mkdir(targetDir, { recursive: true });

    const targetStat = await fs.lstat(target).catch(() => null);
    if (targetStat) {
      if (targetStat.isSymbolicLink()) {
        await fs.unlink(target);
      } else if (targetStat.isDirectory()) {
        await fs.rm(target, { recursive: true, force: true });
      } else {
        await fs.unlink(target);
      }
    }

    const resolvedSource = path.resolve(source);
    const symlinkType = sourceStat.isDirectory() ? "dir" : "file";
    await fs.symlink(resolvedSource, target, symlinkType);

    return true;
  } catch (error) {
    throw new SymlinkError(
      `Failed to symlink ${source} to ${target}: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}
